worker/src/lib.rs

```diff
@@ -0,0 +1,255 @@
+use std::{
+    io,
+    process::{ExitStatus, Stdio},
+    sync::Arc,
+};
+
+use bytes::Bytes;
+use futures::{future::FusedFuture, FutureExt, Stream};
+use log::error;
+use tokio::{
+    io::BufReader,
+    process::{Child, Command},
+    sync::{oneshot, Mutex, Notify, RwLock},
+};
+
+mod logs;
+mod ops;
+
+async fn process_task(
+    mut child: Child,
+    inner: Arc<ProcessInner>,
+    stop_receiver: oneshot::Receiver<()>,
+) {
+    // A child process might close both stdout and stderr,
+    // but remain alive. In that case, we must still try to wait
+    // for the stop message.
+    // Fuse the future to be able to .await twice safely (when !is_terminated()).
+    let mut stop_receiver = stop_receiver.fuse();
+
+    let stdout = BufReader::new(child.stdout.take().expect("should always be available"));
+    let stderr = BufReader::new(child.stderr.take().expect("should always be available"));
+
+    // Phase 1: copy logs from stdout/stderr, on stop message: signal the child.
+    tokio::select! {
+        copied = futures::future::join(
+            logs::copy(stdout, inner.clone()),
+            logs::copy(stderr, inner.clone())
+        ) => {
+            if let Err(e) = copied.0 {
+                error!("{:?}", e);
+            }
+            if let Err(e) = copied.1 {
+                error!("{:?}", e);
+            }
+        },
+        _ = &mut stop_receiver => {
+            if let Err(e) = ops::stop_child(&mut child).await {
+                error!("{:?}", e);
+            }
+        },
+    };
+
+    // Phase 2: stdout/stderr have been closed,
+    // wait for a stop message to arrive (if not arrived yet),
+    // or on the child to finish otherwise
+    loop {
+        tokio::select! {
+            _ = &mut stop_receiver, if !stop_receiver.is_terminated() => {
+                if let Err(e) = ops::stop_child(&mut child).await {
+                    error!("{:?}", e);
+                }
+            },
+            res = child.wait() => {
+                match res {
+                    Ok(s) => {
+                        inner.finish(s).await;
+                        inner.progress.notify_waiters();
+                        return;
+                    }
+                    Err(_) => unimplemented!("when does this happen?"), // TODO
```

Comment:
According to the [manpage](https://linux.die.net/man/3/waitpid), these are the ways that `waitpid()` can fail:
* ECHILD - I'm assuming this is handled by the standard library (it stores a valid child PID internally)
* EINTR - std seems to auto-retry the call (see [[1]](https://github.com/rust-lang/rust/blob/107896c32d5dda4db508968ff34997a39d286966/library/std/src/sys/unix/process/process_unix.rs#L460), [[2]](https://github.com/rust-lang/rust/blob/107896c32d5dda4db508968ff34997a39d286966/library/std/src/sys/unix/mod.rs#L212-L223))
* EINVAL - again, I assume std is always using the correct options for the call.

It seems reasonable to me to panic here, since we do not expect any error and I'm not sure of a nice way to handle an unexpected one.

Reply:
Looked through `std` sources, seems reasonable to panic here. Perhaps using `panic!()` or `unreachable!()` instead?